contracts/staking/StakingPool.sol

```diff
@@ -0,0 +1,484 @@
+        return _isStakingPeriodComplete();
+    }
+
+    function _initializeRewardTokens(
+        address treasury,
+        StakingPoolLib.RewardToken[] calldata _rewardTokens
+    ) internal {
+        for (uint256 i = 0; i < _rewardTokens.length; i++) {
+            IERC20 token = IERC20(_rewardTokens[i].token);
+
+            emit RewardsInitialized(
+                address(token),
+                _rewardTokens[i].totalTokenRewardsAvailable
+            );
+
+            require(
+                token.allowance(treasury, address(this)) >=
+                    _rewardTokens[i].totalTokenRewardsAvailable,
+                "StakingPool: invalid allowance"
+            );
+
+            require(
+                token.transferFrom(
+                    treasury,
+                    address(this),
+                    _rewardTokens[i].totalTokenRewardsAvailable
+                ),
+                "StakingPool: fund tx failed"
+            );
+        }
+    }
+
+    function _withdrawWithoutRewards() internal {
+        User memory user = _users[_msgSender()];
+        require(user.depositAmount > 0, "StakingPool: not eligible");
+
+        delete _users[_msgSender()];
+        StakingPoolLib.Data memory _info = _stakingPoolInfo;
+        _transferStake(uint256((user.depositAmount)), IERC20(_info.stakeToken));
+    }
+
+    function _setRewardsAvailableTimestamp(uint32 timestamp) internal {
+        require(!_isStakingPeriodComplete(), "StakePool: already finalized");
+        //slither-disable-next-line timestamp
+        require(timestamp > block.timestamp, "StakePool: future rewards");
+
+        _stakingPoolInfo.rewardsAvailableTimestamp = timestamp;
+        emit RewardsAvailableTimestamp(timestamp);
+    }
+
+    function _transferStake(uint256 amount, IERC20 stakeToken) internal {
+        emit WithdrawStake(_msgSender(), amount);
+        require(
+            stakeToken.transfer(msg.sender, amount),
+            "StakingPool: stake tx fail"
+        );
+    }
+
+    function _transferRewards(uint256 amount, IERC20 rewardsToken) internal {
+        emit WithdrawRewards(_msgSender(), address(rewardsToken), amount);
+        //slither-disable-next-line calls-loop
+        require(
+            rewardsToken.transfer(_msgSender(), amount),
+            "StakingPool: rewards tx failed"
+        );
+    }
+
+    function _adminEmergencyRewardSweep() internal {
+        StakingPoolLib.RewardToken[] memory rewards = _stakingPoolInfo
+            .rewardTokens;
+        address treasury = _stakingPoolInfo.treasury;
+
+        for (uint256 i = 0; i < rewards.length; i++) {
+            IERC20 token = IERC20(rewards[i].token);
+            require(
+                token.transfer(treasury, rewards[i].totalTokenRewardsAvailable),
+                "StakingPool: withdraw tx failed"
+            );
+        }
+    }
+
+    // fixed
+    function _computeFixedRewards(uint128 amount, uint256 rewardTokenIndex)
```

Comment:
unhappy with these functions `_computeFixedRewards` and `_computeFloatingRewardsPerShare` should be unified in a refactor that has a common _calculate_ function that switches according to `poolType`